File: lib/src/state/settings_by_monitor.ts
import { Rect } from '../utils';
import { SeelenWallWallpaper } from './settings';

export class FancyToolbarSettingsByMonitor {
  enabled: boolean = true;
}

export class SeelenWegSettingsByMonitor {
  enabled: boolean = true;
}

export class WindowManagerSettingsByMonitor {
  enabled: boolean = true;
  padding: number | null = null;
  margin: Rect | null = null;
  gap: number | null = null;
  layout: string | null = null;
}

export class SeelenWallSettingsByMonitor {
  enabled: boolean = true;
  backgrounds: SeelenWallWallpaper[] | null = null;
}

export enum WorkspaceIdentifierType {
  Name = 'name',
  Index = 'index',
}

export class WorkspaceIdentifier {
  id: string;
  kind: WorkspaceIdentifierType;

  constructor(id: string, kind: WorkspaceIdentifierType) {
    this.id = id;
    this.kind = kind;
  }
}

export class WorkspaceConfiguration {
  identifier: WorkspaceIdentifier;
  layout: string | null = null;
  backgrounds: SeelenWallWallpaper[] | null = null;

  constructor(identifier: WorkspaceIdentifier) {
    this.identifier = identifier;
  }
}

export class MonitorConfiguration {
  tb: FancyToolbarSettingsByMonitor = new FancyToolbarSettingsByMonitor();
  wall: SeelenWallSettingsByMonitor = new SeelenWallSettingsByMonitor();
  weg: SeelenWegSettingsByMonitor = new SeelenWegSettingsByMonitor();
  wm: WindowManagerSettingsByMonitor = new WindowManagerSettingsByMonitor();
  /** list of settings by workspace on this monitor */
  workspaces: WorkspaceConfiguration[] = [];
}

File: lib/src/state/settings_by_monitor.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::rect::Rect;

use super::SeelenWallWallpaper;

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(default, rename_all = "camelCase")]
pub struct FancyToolbarSettingsByMonitor {
    pub enabled: bool,
}

impl Default for FancyToolbarSettingsByMonitor {
    fn default() -> Self {
        Self { enabled: true }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(default, rename_all = "camelCase")]
pub struct SeelenWegSettingsByMonitor {
    pub enabled: bool,
}

impl Default for SeelenWegSettingsByMonitor {
    fn default() -> Self {
        Self { enabled: true }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(default, rename_all = "camelCase")]
pub struct WindowManagerSettingsByMonitor {
    pub enabled: bool,
    pub padding: Option<u32>,
    pub margin: Option<Rect>,
    pub gap: Option<u32>,
    pub layout: Option<String>,
}

impl Default for WindowManagerSettingsByMonitor {
    fn default() -> Self {
        Self {
            enabled: true,
            padding: None,
            margin: None,
            gap: None,
            layout: None,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(default, rename_all = "camelCase")]
pub struct SeelenWallSettingsByMonitor {
    pub enabled: bool,
    pub backgrounds: Option<Vec<SeelenWallWallpaper>>,
}

impl Default for SeelenWallSettingsByMonitor {
    fn default() -> Self {
        Self {
            enabled: true,
            backgrounds: None,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
pub enum WorkspaceIdentifierType {
    #[serde(alias = "name")]
    Name,
    #[serde(alias = "index")]
    Index,
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceIdentifier {
    pub id: String,
    pub kind: WorkspaceIdentifierType,
}

#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceConfiguration {
    pub identifier: WorkspaceIdentifier,
    pub layout: Option<String>,
    pub backgrounds: Option<Vec<SeelenWallWallpaper>>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, JsonSchema)]
#[serde(default, rename_all = "camelCase")]
pub struct MonitorConfiguration {
    pub tb: FancyToolbarSettingsByMonitor,
    pub weg: SeelenWegSettingsByMonitor,
    pub wm: WindowManagerSettingsByMonitor,
    pub wall: SeelenWallSettingsByMonitor,
    /// list of settings by workspace on this monitor
    pub workspaces_v2: Vec<WorkspaceConfiguration>,
}
